src/libpriv/rpmostree-core.c

```diff
@@ -2515,50 +2646,39 @@ rpmostree_context_assemble_tmprootfs (RpmOstreeContext      *self,
                                        cancellable, error))
         return FALSE;
     }
-  else
-    {
-      /* Otherwise, we unpack the first package to get the initial
-       * rootfs dir.
-       */
-      rpmte te = rpmtsElement (ordering_ts, 0);
-      DnfPackage *pkg = (void*)rpmteKey (te);
-
-      g_assert (pkg);
-
-      if (!checkout_package_into_root (self, pkg,
-                                       tmprootfs_dfd, ".", devino_cache,
-                                       g_hash_table_lookup (pkg_to_ostree_commit,
-                                                            pkg),
-                                       cancellable, error))
-        return FALSE;
-
-      filesystem_package = g_object_ref (pkg);
-    }
 
   for (guint i = 0; i < n_rpmts_elements; i++)
     {
       rpmte te = rpmtsElement (ordering_ts, i);
-      DnfPackage *pkg = (void*)rpmteKey (te);
-
-      g_assert (pkg);
+      rpmElementType type = rpmteType (te);
 
-      if (pkg == filesystem_package)
-        continue;
+      if (type == TR_ADDED)
+        {
+          DnfPackage *pkg = (void*)rpmteKey (te);
+          if (pkg == filesystem_package)
+            continue;
 
-      if (!checkout_package_into_root (self, pkg,
-                                       tmprootfs_dfd, ".", devino_cache,
-                                       g_hash_table_lookup (pkg_to_ostree_commit,
-                                                            pkg),
-                                       cancellable, error))
-        return FALSE;
+          if (!checkout_package_into_root (self, pkg, tmprootfs_dfd, ".", devino_cache,
+                                           g_hash_table_lookup (pkg_to_ostree_commit, pkg),
+                                           cancellable, error))
+            return FALSE;
+        }
+      else
+        {
+          g_assert (type == TR_REMOVED);
+          if (!delete_package_from_root (self, te, tmprootfs_dfd, cancellable, error))
+            return FALSE;
+        }
     }
 
   rpmostree_output_task_end ("done");
 
   if (!rpmostree_rootfs_prepare_links (tmprootfs_dfd, cancellable, error))
     return FALSE;
 
-  if (!noscripts)
+  /* NB: we're not running scripts right now for removals, so this is only for
+   * overlays */
```

Comment:
That is correct! I did have a look when writing this code and indeed the great majority was systemd, ldconfig, and icon cache.

Hehe, good point about `NB`. I'm actually not sure myself how I decide when to preface by `NB` and when not to. I think maybe when something is especially tricky?